### PR TITLE
Switch to Boost generic sockets

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Boost has generic sockets as an abstraction between TCP and UNIX domain sockets. This helps avoid `#ifdef`s and `template`s around the codebase as well as making it easier to work with different socket types in a generic manner.

## Related issue number

#631

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/common/client_connection.h [hunk 1]
```diff
@@ -15,24 +15,27 @@
...
-#include <deque>
-#include <memory>
-
-#include <boost/asio.hpp>
+#include <boost/asio/basic_stream_socket.hpp>
+#include <boost/asio/buffer.hpp>
 #include <boost/asio/error.hpp>
+#include <boost/asio/generic/stream_protocol.hpp>
 #include <boost/enable_shared_from_this.hpp>
+#include <deque>
```
Comment (pcmoritz) [inline src/ray/common/client_connection.h:23]: C++ standard headers should still come first, before the boost ones

File: src/ray/raylet/client_connection_test.cc [hunk 2]
```diff
@@ -35,18 +35,21 @@ class ClientConnectionTest : public ::testing::Test {
   ClientConnectionTest()
       : io_service_(), in_(io_service_), out_(io_service_), error_message_type_(1) {
 #if defined(BOOST_ASIO_HAS_LOCAL_SOCKETS)
-    boost::asio::local::connect_pair(in_, out_);
+    boost::asio::local::stream_protocol::socket input(io_service_), output(io_service_);
```
Comment (pcmoritz) [inline src/ray/raylet/client_connection_test.cc:38]: Hmm, why is this change needed

File: src/ray/util/url.cc [hunk 3]
```diff
@@ -16,64 +16,103 @@
...
+#endif
+
 #include <algorithm>
-#include <boost/asio/ip/address.hpp>
+#include <boost/asio/generic/stream_protocol.hpp>
+#ifndef _WIN32
+#include <boost/asio/local/stream_protocol.hpp>
+#endif
 #include <boost/asio/ip/tcp.hpp>
+#include <sstream>
```
Comment (pcmoritz) [inline src/ray/util/url.cc:31]: Same as above, the headers should be moved up (see https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes). I realize this was not the case before, but we should fix it.

File: src/ray/util/url.cc [hunk 4]
```diff
@@ -16,64 +16,103 @@
...
-  size_t port_end = endpoint.find('/', host_begin);
-  if (port_end >= endpoint.size()) {
-    port_end = endpoint.size();
+namespace ray {
+
+std::string endpoint_to_url(
+    const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol> &ep,
+    bool include_scheme) {
+  std::string result;
+  switch (ep.protocol().family()) {
```
Comment (pcmoritz) [inline src/ray/util/url.cc:44]: For this whole block we can just call `ep.address().to_string()`, like this:
```
if (ep.protocol().family() == AF_INET || ep.protocol().family() == AF_INET6) {
    return "tcp://" + ep.address().to_string();
}
```

I'd say let's get rid of the `include_schema` option and always include it in the `endpoint_to_url`, there is not really a reason to not include it, is there?

File: src/ray/util/url.cc [hunk 5]
```diff
@@ -16,64 +16,103 @@
...
-      if (port_begin < port_end) {
-        ++port_begin;
-      } else {
-        port_begin = port_end;
-      }
-    } else {
-      host_end = port_end;
-      port_begin = host_end;
+  case AF_UNIX: {
+    size_t path_offset = offsetof(sockaddr_un, sun_path);
```
Comment (pcmoritz) [inline src/ray/util/url.cc:62]: I feel like the above approach of `ep.address().to_string()` should also work for this code path, can you check? In that case they can be unified via
```
std::string address = ep.address().to_string();
if (ep.protocol().family() == AF_INET || ep.protocol().family() == AF_INET6) {
    return "tcp://" + address;
} else if (ep.protocol().family() == AF_UNIX) {
  return "unix:///" + address
} else {
  RAY_LOG(FATAL) << "unsupported protocol family: " << ep.protocol().family();
}
```

File: src/ray/common/client_connection.h (hunk 1 above)
Comment (mehrdadn) [reply src/ray/common/client_connection.h:23]: I'll give it a shot, but I seem to recall the linter complaining when I've done that in the past.

File: src/ray/util/url.cc [hunk 6]
```diff
@@ -16,64 +16,103 @@
...
+  }
+  default:
+    RAY_LOG(FATAL) << "unsupported protocol family: " << ep.protocol().family();
+    break;
+  }
+  return result;
+}
+
+boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>
+parse_url_endpoint(const std::string &endpoint, int default_port) {
```
Comment (pcmoritz) [inline src/ray/util/url.cc:78]: This method should be simplified a bunch (a little pseudocode-y):
```
std::string schema;
std::string url;
if (endpoint.startswith("unix://")) {
  schema = "unix";
  url = endpoint.lstrip("unix://")
} else if (endpoint.startswith("/")) {
  // I would only support absolute unix paths.
  schema = "unix";
  url = endpoint;
} else if (endpoint.startswith("tcp://")) {
  schema = "tcp";
  url = endpoint.lstrip("tcp://");
} else {
  schema = "tcp";
  url = endpoint;
}

if (scheme == "unix") {
  // I don't think we need the windows ifdef here? This code path should not be called on windows, right?
  return boost::asio::local::stream_protocol::endpoint(url);
}
if (scheme == "tcp") {
  // The ScanToken stuff, but let's make ScanToken a private method and not encourage use outside of this file.
  // Also let's use std::stoi for the port parsing.
}
```

File: src/ray/raylet/client_connection_test.cc (hunk 2 above)
Comment (mehrdadn) [reply src/ray/raylet/client_connection_test.cc:38]: You can't `connect_pair` generic sockets, it's a `boost::asio::local` thing. (It doesn't compile otherwise.)

File: src/ray/util/url.cc (hunk 5 above)
Comment (mehrdadn) [reply src/ray/util/url.cc:62]: [There's no `address()` method for generic sockets](https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/generic__basic_endpoint.html), only for specialized socket types (TCP, local), so it won't work here unfortunately.

File: src/ray/util/url.cc (hunk 4 above)
Comment (mehrdadn) [reply src/ray/util/url.cc:44]: Well the reason to not always include it is that `ClientConnection::RemoteEndpointInfo()` didn't include it before, and I wasn't trying to change its behavior. Should I?

Comment (mehrdadn) [reply src/ray/util/url.cc:44]: Regarding `address()`, it won't work; see my reply to the other comment below.

File: src/ray/util/url.cc (hunk 6 above)
Comment (mehrdadn) [reply src/ray/util/url.cc:78]: Done, but note that this won't support `scheme:path` anymore, only `scheme://path`. [It seems both are valid](https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes?), so I tried to handle both properly—that's part of the reason why it was a bit more complicated than just stripping a prefix.

File: src/ray/common/client_connection.h (hunk 1 above)
Comment (mehrdadn) [reply src/ray/common/client_connection.h]: Done

File: src/ray/util/url.cc (hunk 3 above)
Comment (mehrdadn) [reply src/ray/util/url.cc]: Done

File: src/ray/util/url.cc [hunk 7]
```diff
@@ -16,64 +16,114 @@
...
+  std::string result;
+  switch (ep.protocol().family()) {
+  case AF_INET:
+  case AF_INET6: {
+    boost::asio::ip::tcp protocol = ep.protocol().family() == AF_INET
+                                        ? boost::asio::ip::tcp::v4()
+                                        : boost::asio::ip::tcp::v6();
+    boost::asio::ip::tcp::endpoint e(protocol, 0);
+    RAY_DCHECK(e.size() == ep.size());
+    memcpy(e.data(), ep.data(), ep.size());
```
Comment (pcmoritz) [inline src/ray/util/url.cc:68]: It seems to me that this line makes some assumption about the representation of `boost::asio::ip::tcp::endpoint` and `boost::asio::generic::basic_endpoint`, seems unsafe. There is an explicit way to convert from the tcp endpoint to the basic endpoint in the boost documentation, but I don't think the other way around is supported. Is that right?

File: src/ray/util/url.cc [hunk 8]
```diff
@@ -16,64 +16,114 @@
...
-        ++port_begin;
-      } else {
-        port_begin = port_end;
-      }
-    } else {
-      host_end = port_end;
-      port_begin = host_end;
+  case AF_UNIX: {
+    size_t path_offset = offsetof(sockaddr_un, sun_path);
+    result.append(reinterpret_cast<const char *>(ep.data()) + path_offset,
```
Comment (pcmoritz) [inline src/ray/util/url.cc:79]: Same comment as above.

File: src/ray/util/url.cc [hunk 9]
```diff
@@ -16,64 +16,114 @@
...
+
+boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>
+parse_url_endpoint(const std::string &endpoint, int default_port) {
+  // Syntax reference: https://en.wikipedia.org/wiki/URL#Syntax
+  // Note that we're a bit more flexible, to allow parsing "127.0.0.1" as a URL.
+  boost::asio::generic::stream_protocol::endpoint result;
+  std::string address = endpoint, scheme;
+  if (address.find("unix://") == 0) {
+    scheme = "unix";
+    address.erase(0, scheme.size() + 3);
```
Comment (pcmoritz) [inline src/ray/util/url.cc:101]: To avoid this arithmetic, let's just user `unix://` as the scheme, same below.

File: src/ray/util/url.cc (hunk 7 above)
Comment (mehrdadn) [reply src/ray/util/url.cc:68]: That's correct regarding the conversion. However, this is just the C++ version of the mandatory cast from `struct sockaddr *` to `struct sockddr_in *` (or `struct sockaddr_un *`) in C, which is the way you're supposed to do it. So I guess it's "implicit" in some sense, but not in the sense that I'd expect it to randomly break with a Boost upgrade.

File: src/ray/util/url.cc (hunk 8 above)
Comment (mehrdadn) [reply src/ray/util/url.cc:79]: Yup replied above.

File: src/ray/util/url.cc (hunk 7 above)
Comment (pcmoritz) [reply src/ray/util/url.cc:68]: After digging a bit more into this, it seems that you can use this assignment operator to do it in a nicer way:
```
https://github.com/chriskohlhoff/asio/blob/8087252a0c3c2f0baad96ddbd6554db17a846376/asio/include/asio/generic/basic_endpoint.hpp#L98
```

Also,  instead of using the stringstreams below, we can do `result = e.address()`, right?

Comment (mehrdadn) [reply src/ray/util/url.cc:68]: I think that's only assigning an instance of the same endpoint class, not a different one?

Comment (mehrdadn) [reply src/ray/util/url.cc:68]: When you're inside `template <typename Protocol> class basic_endpoint`, `basic_endpoint` by itself just refers to `basic_endpoint<Protocol>`. You'd want `basic_endpoint<Other>` or something like that here for this to work.

File: src/ray/util/url.cc (hunk 9 above)
Comment (mehrdadn) [reply src/ray/util/url.cc]: Pushed

File: src/ray/util/url.cc (hunk 7 above)
Comment (pcmoritz) [reply src/ray/util/url.cc:68]: Then maybe this: https://github.com/chriskohlhoff/asio/blob/8087252a0c3c2f0baad96ddbd6554db17a846376/asio/include/asio/generic/basic_endpoint.hpp#L70? There must be a way to do this without mem copy...

Comment (mehrdadn) [reply src/ray/util/url.cc:68]: Oh interesting, that's not inside the documentation. Should I use it anyway? It might break later...

Comment (mehrdadn) [reply src/ray/util/url.cc:68]: Actually I don't think that's working? In fact now I remember I ran into this compile error when I tried this myself:

	src/ray/util/url.cc(63,36): error: no matching constructor for initialization of 'boost::asio::ip::tcp::endpoint' (aka 'basic_endpoint<boost::asio::ip::tcp>')
	    boost::asio::ip::tcp::endpoint e(ep);
	                                   ^ ~~
	external/boost\boost/asio/ip/basic_endpoint.hpp(97,3): note: candidate constructor not viable: no known conversion from 'const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>' to 'const boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>' for 1st argument
	  basic_endpoint(const basic_endpoint& other) BOOST_ASIO_NOEXCEPT
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(104,3): note: candidate constructor not viable: no known conversion from 'const boost::asio::generic::basic_endpoint<boost::asio::generic::stream_protocol>' to 'boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>' for 1st argument
	  basic_endpoint(basic_endpoint&& other) BOOST_ASIO_NOEXCEPT
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(60,3): note: candidate constructor not viable: requires 0 arguments, but 1 was provided
	  basic_endpoint() BOOST_ASIO_NOEXCEPT
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(81,3): note: candidate constructor not viable: requires 2 arguments, but 1 was provided
	  basic_endpoint(const InternetProtocol& internet_protocol,
	  ^
	external/boost\boost/asio/ip/basic_endpoint.hpp(90,3): note: candidate constructor not viable: requires 2 arguments, but 1 was provided
	  basic_endpoint(const boost::asio::ip::address& addr,
	  ^

**Review by pcmoritz — APPROVED** [review]
